lib/JSON/Validator.pm

```diff
@@ -573,69 +573,85 @@ sub _validate {
 
 sub _validate_all_of {
   my ($self, $data, $path, $rules) = @_;
-  my $type = data_type $data, $rules;
-  my (@errors, @expected);
+  my (@errors);
 
   my $i = 0;
   for my $rule (@$rules) {
     next unless my @e = $self->_validate($_[1], $path, $rule);
-    my $schema_type = schema_type $rule;
-    push @expected, $schema_type if $schema_type;
-    push @errors, [$i, @e] if !$schema_type or $schema_type eq $type;
+    push @errors, [$i, @e];
   }
   continue {
     $i++;
   }
 
-  return E $path, [allOf => type => join('/', uniq @expected), $type]
-    if !@errors and @expected;
-  return prefix_errors allOf => @errors if @errors;
-  return;
+  return if not @errors;
+
+  return E $path,
+    [
+    allOf => type => join('/', uniq map $_->[1]->details->[0], @errors),
+    $errors[-1][1]->details->[2]
+    ]
+    if @errors > 1 and not grep $_->[1]->details->[1] ne 'type', @errors;
+
+  return prefix_errors allOf => @errors;
 }
 
 sub _validate_any_of {
-  my ($self, $data, $path, $rules) = @_;
-  my $type = data_type $data, $rules;
-  my (@e, @errors, @expected);
+  my ($self, $data, $path, $rules, $prefix) = @_;
+  $prefix //= 1;
+  my (@e, @errors);
 
   my $i = 0;
   for my $rule (@$rules) {
     @e = $self->_validate($_[1], $path, $rule);
+
+    # TODO: the spec might not think that short-circuiting is ok here
```

Comment:
What do you mean with this comment? Why would you use CPU cycles checking for more errors, if it already fulfils anyOf?

Reply:
The draft 2019-09 spec has a few things to say about verifying all branches of a conditional even if some parts already gave a sufficient answer to report up to the next level.  I didn't mean to commit that comment; it was a note to myself to go back and read the spec more carefully.